### PR TITLE
feat(AWS CloudFront): Support `behavior.CachePolicyId`

### PR DESCRIPTION
**Pull request by issea1015**
Closes: #9606

### CODE REVIEW

File: docs/providers/aws/events/cloudfront.md [hunk 1]
```diff
@@ -238,6 +238,20 @@ functions:
             id: 658327ea-f89d-4fab-a63d-7e88639e58f6 # references AWS Managed Policy named Managed-CachingOptimized
 ```
 
+You also can do the same thing with below configuration. But the `cloudFront.behavior.CachePolicyId` will be overwritten by `cloudFront.cachePolicy`
```
Comment (pgrzesik) [inline docs/providers/aws/events/cloudfront.md:241]: Let's rephrase this as:

```
It is also possible to reference policies with `behavior.CachePolicyId` property. When both `cachePolicy.id` and `behavior.CachePolicyId` are specified, setting from `cachePolicy.id` will be used.
```

File: lib/plugins/aws/package/compile/events/cloudFront.js [hunk 2]
```diff
@@ -432,6 +432,13 @@ class AwsCompileCloudFrontEvents {
             };
             let shouldAssignCachePolicy = true;
             if (event.cloudFront.behavior) {
+              if (event.cloudFront.behavior.CachePolicyId) {
+                if (!event.cloudFront.cachePolicy) {
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/events/cloudFront.js:436]: Let's have that condition under a single if instead of two

File: lib/plugins/aws/package/compile/events/cloudFront.js [hunk 3]
```diff
@@ -432,6 +432,13 @@ class AwsCompileCloudFrontEvents {
...
+                  event.cloudFront.cachePolicy = {
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/events/cloudFront.js:437]: Instead of mutating configuration of the event we should set that property directly on `behavior` object that will be passed to CloudFormation resource

File: test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js [hunk 4]
```diff
@@ -1996,6 +1997,39 @@ describe('test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js',
                 },
               ],
             },
+            fnCachePolicyManagedAlternativeWay: {
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js:2000]: Maybe let's use more descriptive name instead of using suffix `AlternativeWay`? e.g. `fnCachePolicyManagedSetViaBehavior`

File: test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js [hunk 5]
```diff
@@ -2113,6 +2147,18 @@ describe('test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js',
       expect(getAssociatedCacheBehavior('managedPolicy').CachePolicyId).to.eq(cachePolicyId);
     });
 
+    it('Should attach a cache policy to a cloudfront behavior when specified by id in alternative way of lambda config', () => {
+      expect(getAssociatedCacheBehavior('managedPolicyInAlternativeWay').CachePolicyId).to.eq(
+        cachePolicyId
+      );
+    });
+
+    it('Should attach a cache policy specified in major way to a cloudfront behavior when specified by id in both ways of lambda config', () => {
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js:2156]: Let's point out directly how it was set instead of using ambiguous `major way` statement

File: test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js [hunk 6]
```diff
@@ -2113,6 +2147,18 @@ describe('test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js',
       expect(getAssociatedCacheBehavior('managedPolicy').CachePolicyId).to.eq(cachePolicyId);
     });
 
+    it('Should attach a cache policy to a cloudfront behavior when specified by id in alternative way of lambda config', () => {
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/package/compile/events/cloudFront.test.js:2150]: Let's point out directly how it was set instead of using ambiguous `alternative way` statement

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @issea1015 :bow: It looks good in general, I have a few minor comments, please let me know what do you think :bow:

**Comment by issea1015** [comment]
@pgrzesik Thanks a lot for the detailed feedback and your time! I fully agree with you. Let me update the PR and let you know :)

File: lib/plugins/aws/package/compile/events/cloudFront.js [hunk 7]
```diff
@@ -444,16 +444,28 @@ class AwsCompileCloudFrontEvents {
...
+            }
+
+            if (event.cloudFront.behavior && event.cloudFront.behavior.CachePolicyId) {
+              Object.assign(behavior, {
+                CachePolicyId: event.cloudFront.behavior.CachePolicyId,
+              });
+              shouldAssignCachePolicy = false;
+            }
+
             if (event.cloudFront.cachePolicy) {
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/events/cloudFront.js:468]: If I'm not mistaken, in the current logic, if the `cachePolicy.name` is specified, it will also override `behavior.CachePolicyId` setting, correct? I think it would be good to document it as well

**Review by pgrzesik — COMMENTED** [review]
Thank you @issea1015 - I have only one question left and we'll be good to go :raised_hands:

File: lib/plugins/aws/package/compile/events/cloudFront.js (hunk 7 above)
Comment (issea1015) [reply lib/plugins/aws/package/compile/events/cloudFront.js:468]: @pgrzesik Thanks for the quick reply 👍 Yeah, that's correct. Let me improve PR and let you know :)

Comment (issea1015) [reply lib/plugins/aws/package/compile/events/cloudFront.js:468]: @pgrzesik I just pushed a change. I considered adding some more tests for this `name` property case. In the end, I thought that it can be redundant to add more tests for it as the `name` case would have almost the same call stack trace as the `id` case. Could you let me know what do you think? :bow:

And not sure I updated the document in a well-aligned way to the community convention. Please feel free to let me know how I can polish the phrases :bow:

File: docs/providers/aws/events/cloudfront.md [hunk 8]
```diff
@@ -238,6 +238,20 @@ functions:
             id: 658327ea-f89d-4fab-a63d-7e88639e58f6 # references AWS Managed Policy named Managed-CachingOptimized
 ```
 
+It is also possible to reference policies with `behavior.CachePolicyId` property. When both `cachePolicy.(id|name)` and `behavior.CachePolicyId` are specified, setting from `cachePolicy.(id|name)` will be used.
```
Comment (pgrzesik) [inline docs/providers/aws/events/cloudfront.md:241]: I think instead of presenting it this way, we could add a sentence like this. 

```
Similarily, when `cachePolicy.name` and `behavior.CachePolicyId` are specified, setting from `cachePolicy.name` will be used. 
```

The reason for being more descriptive here is that I don't think we use similar notation in our documentation anywhere and I think in such cases it's better to be a little bit more verbose. Let me know what do you think :bow:

**Review by pgrzesik — CHANGES_REQUESTED** [review]

File: lib/plugins/aws/package/compile/events/cloudFront.js (hunk 7 above)
Comment (pgrzesik) [reply lib/plugins/aws/package/compile/events/cloudFront.js:468]: I think it might be nice to cover it with a single test where `cachePolicy.name` takes precedence over `behavior.CachePolicyId`.

File: docs/providers/aws/events/cloudfront.md (hunk 8 above)
Comment (issea1015) [reply docs/providers/aws/events/cloudfront.md:241]: Fully agree with this. Thanks for the guidance 🙇  I will update the PR and let you know.

File: lib/plugins/aws/package/compile/events/cloudFront.js (hunk 7 above)
Comment (issea1015) [reply lib/plugins/aws/package/compile/events/cloudFront.js:468]: That will be more solid. Thanks for the guidance 🙇 I will update the PR and let you know.

**Review by pgrzesik — APPROVED** [review]
Well done, thanks a lot for your contribution @issea1015 :bow:

**Comment by issea1015** [comment]
@pgrzesik I've enjoyed this first journey in this repository with your full support! Thanks a lot 🙌

**Comment by pgrzesik** [comment]
@issea1015 I'm really happy to hear that, thanks for your great contribution :raised_hands:
